### ISSUE TITLE
SAM-cls

### ISSUE THREAD

**Issue by gykj-hmf** [KyanChen/RSPrompter#26]
How use resnet18 (trained with label) to classify？In your project, I didn't find the config file <'configs/ins_seg/samcls_{data_set_name}_config.py'>

**Comment by KyanChen** [comment]
1. crop and label all the instances with mask generated by SAM's "everthing" mode. https://github.com/KyanChen/RSPrompter/blob/cky/tools/ins_seg/sam/sam_cls/get_sam_cls_crops.py
2. train the resnet 18
3. get the classification metrics. https://github.com/KyanChen/RSPrompter/blob/cky/tools/ins_seg/sam/sam_cls/get_sam_cls_metrics.py

```
# custom_imports = dict(imports=['mmseg.datasets', 'mmseg.models', 'mmdet.models'], allow_failed_imports=False)

max_epochs = 500

optimizer = dict(
    type='AdamW',
    lr=0.0005,
    weight_decay=1e-4
)

param_scheduler = [
    # warm up learning rate scheduler
    dict(
        type='LinearLR',
        start_factor=5e-4,
        by_epoch=True,
        begin=0,
        end=1,
        # update by iter
        convert_to_iter_based=True),
    # main learning rate scheduler
    dict(
        type='CosineAnnealingLR',
        T_max=max_epochs,
        by_epoch=True,
        begin=1,
        end=max_epochs,
    )
]

param_scheduler_callback = dict(
    type='ParamSchedulerHook'
)

num_classes = 10

evaluator_ = dict(
        type='Accuracy',
        task="multiclass",
        num_classes=num_classes+1,
)

evaluator = dict(
    # train_evaluator=evaluator_,
    val_evaluator=evaluator_,
    test_evaluator=evaluator_
)


data_preprocessor = dict(
    num_classes=num_classes+1,
    # RGB format normalization parameters
    mean=[123.675, 116.28, 103.53],
    std=[58.395, 57.12, 57.375],
    # convert image from BGR to RGB
    to_rgb=True,
)

model = dict(
    type='mmpretrain.ImageClassifier',
    data_preprocessor=data_preprocessor,
    backbone=dict(
        type='mmpretrain.ResNet',
        depth=18,
        num_stages=4,
        out_indices=(3, ),
        style='pytorch'),
    neck=dict(type='mmpretrain.GlobalAveragePooling'),
    head=dict(
        type='mmpretrain.LinearClsHead',
        num_classes=num_classes+1,
        in_channels=512,
        loss=dict(type='mmpretrain.CrossEntropyLoss', loss_weight=1.0),
        topk=(1, 5),
    ),
    pretrained='https://download.openmmlab.com/mmclassification/v0/resnet/resnet18_8xb32_in1k_20210831-fbbb1da6.pth'
)

model_cfg = dict(
    type='MMClsPLer',
    hyperparameters=dict(
        optimizer=optimizer,
        param_scheduler=param_scheduler,
        evaluator=evaluator,
    ),
    whole_model=model,
)

task_name = 'nwpu_ins'
exp_name = 'E20230601_10'
logger = dict(
    type='WandbLogger',
    project=task_name,
    group='samcls',
    name=exp_name
)
# logger = None


callbacks = [
    param_scheduler_callback,
    dict(
        type='ModelCheckpoint',
        dirpath=f'results/{task_name}/{exp_name}/checkpoints',
        save_last=True,
        mode='max',
        monitor='valmulticlassaccuracy_0',
        save_top_k=3,
        filename='epoch_{epoch}-map_{valmulticlassaccuracy_0:.4f}'
        # mode='min',
        # monitor='train_loss',
        # save_top_k=2,
        # filename='epoch_{epoch}-trainloss_{train_loss:.4f}'
    ),
    dict(
        type='LearningRateMonitor',
        logging_interval='step'
    )
]


trainer_cfg = dict(
    compiled_model=False,
    accelerator="auto",
    strategy="auto",
    # strategy="ddp",
    # strategy='ddp_find_unused_parameters_true',
    # precision='32',
    # precision='16-mixed',
    devices=[4, 5, 6, 7],
    default_root_dir=f'results/{task_name}/{exp_name}',
    # default_root_dir='results/tmp',
    max_epochs=max_epochs,
    logger=logger,
    callbacks=callbacks,
    log_every_n_steps=5,
    check_val_every_n_epoch=5,
    benchmark=True,
    # sync_batchnorm=True,
    # fast_dev_run=True,

    # limit_train_batches=1,
    # limit_val_batches=0,
    # limit_test_batches=None,
    # limit_predict_batches=None,
    # overfit_batches=0.0,

    # val_check_interval=None,
    # num_sanity_val_steps=0,
    # enable_checkpointing=None,
    # enable_progress_bar=None,
    # enable_model_summary=None,
    # accumulate_grad_batches=32,
    # gradient_clip_val=15,
    # gradient_clip_algorithm='norm',
    # deterministic=None,
    # inference_mode: bool=True,
    use_distributed_sampler=True,
    # profiler="simple",
    # detect_anomaly=False,
    # barebones=False,
    # plugins=None,
    # reload_dataloaders_every_n_epochs=0,
)


train_pipeline = [
    dict(type='mmpretrain.LoadImageFromFile'),
    dict(type='mmpretrain.RandomResizedCrop', scale=224),
    dict(type='mmpretrain.RandomFlip', prob=0.5, direction='horizontal'),
    dict(type='mmpretrain.PackInputs'),
]

test_pipeline = [
    dict(type='mmpretrain.LoadImageFromFile'),
    dict(type='mmpretrain.ResizeEdge', scale=256, edge='short'),
    dict(type='mmpretrain.CenterCrop', crop_size=224),
    dict(type='mmpretrain.PackInputs'),
]


train_batch_size_per_gpu = 256
train_num_workers = 8
test_batch_size_per_gpu = 256
test_num_workers = 8
persistent_workers = True

dataset_type = 'mmpretrain.ImageNet'
data_root = '/data/kyanchen/cache_data/ins_seg/sam_cls'
train_data_prefix = 'train'
val_data_prefix = 'val'

val_loader = dict(
        batch_size=test_batch_size_per_gpu,
        num_workers=test_num_workers,
        persistent_workers=persistent_workers,
        pin_memory=True,
        dataset=dict(
            type=dataset_type,
            data_root=data_root,
            ann_file='samcls_val.txt',
            data_prefix=val_data_prefix,
            test_mode=True,
            pipeline=test_pipeline))

datamodule_cfg = dict(
    type='PLDataModule',
    train_loader=dict(
        batch_size=train_batch_size_per_gpu,
        num_workers=train_num_workers,
        persistent_workers=persistent_workers,
        pin_memory=True,
        dataset=dict(
            type=dataset_type,
            data_root=data_root,
            ann_file='samcls_train.txt',
            data_prefix=train_data_prefix,
            pipeline=train_pipeline)
    ),
    val_loader=val_loader,
    # predict_loader=val_loader
)
```
